### PR TITLE
Add `insideFunctions: {"function": int}` to `number-max-precision`

### PR DESCRIPTION
**Pull request by romainmenke**
> Which issue, if any, is this issue related to?

Closes https://github.com/stylelint/stylelint/issues/6856

> Is there anything in the PR that needs further explanation?

I've switched this rule to `@csstools/css-parser-algorithms` because the walker state feature is ideal to add support for this new plugin option.

### CODE REVIEW

File: lib/rules/number-max-precision/index.js [hunk 1]
```diff
@@ -72,47 +85,122 @@ const rule = (primary, secondaryOptions) => {
 				return;
 			}
 
-			valueParser(value).walk((valueNode) => {
-				const { unit } = getDimension(valueNode);
+			parseListOfComponentValues(tokenize({ css: value })).forEach((componentValue) => {
+				/** @type {{ ignored: boolean, precision: number }} */
```
Comment (ybiquitous) [inline lib/rules/number-max-precision/index.js:89]: [suggest] When specifying a primary option type like this:

```diff
-/** @type {import('stylelint').Rule} */
+/** @type {import('stylelint').Rule<number>} */
```

This `@type` annotation seems no longer needed:

```suggestion
```

File: lib/rules/number-max-precision/index.js [hunk 2]
```diff
@@ -72,47 +85,122 @@ const rule = (primary, secondaryOptions) => {
...
+		 * @param {string} functionName
+		 * @param {number} currentPrecision
+		 * @returns {number}
+		 */
+		function precisionInsideFunction(functionName, currentPrecision) {
+			if (functionName in secondaryOptions.insideFunctions) {
+				return secondaryOptions.insideFunctions[functionName];
+			}
+
+			for (const [name, precision] of Object.entries(secondaryOptions.insideFunctions)) {
```
Comment (ybiquitous) [inline lib/rules/number-max-precision/index.js:196]: [suggest] Constructing `insideFunctions` before walking might be more readable and have a bit performance benefit. Any thoughts?

```js
if (!validOptions) {
	return;
}

/** @type {Map<string, number>} */
const insideFunctions = new Map(Object.entries(secondaryOptions?.insideFunctions ?? {}));
```

```suggestion
			const precision = insideFunctions.get(functionName);
			if (isNumber(precision)) return precision;

			for (const [name, precision] of insideFunctions) {
```

**Review by ybiquitous — CHANGES_REQUESTED** [review]
@romainmenke Thanks for the pull request! Could you address the reviews?

File: lib/rules/number-max-precision/README.md [hunk 3]
```diff
@@ -133,3 +133,44 @@ a {
   width: 10.989my-other-unit;
 }
 ```
+
+### `insideFunctions: {"/regex/": 1, /regex/: 1, "string": 1}`
```
[the remark this answers is not in the document]
Comment (romainmenke) [reply lib/rules/number-max-precision/README.md:137]: Good catch 🙇

File: lib/rules/number-max-precision/index.js [hunk 4]
```diff
@@ -72,47 +89,123 @@ const rule = (primary, secondaryOptions) => {
...
 			});
 		}
+
+		/**
+		 * @param {string} functionName
+		 * @param {number} currentPrecision
+		 * @returns {number}
+		 */
+		function precisionInsideFunction(functionName, currentPrecision) {
+			{
```
Comment (romainmenke) [inline lib/rules/number-max-precision/index.js:195]: I've added a block here because otherwise `precision` is declared twice.

I don't have a strong preference here.
Changing the names would also work.

File: lib/rules/number-max-precision/index.js (hunk 2 above)
Comment (romainmenke) [reply lib/rules/number-max-precision/index.js]: Yes, much better, thank you

File: lib/rules/number-max-precision/index.js (hunk 4 above)
Comment (ybiquitous) [reply lib/rules/number-max-precision/index.js:195]: Hum, how about the name `precisionForFunction` instead of `precision`?

File: lib/rules/number-max-precision/index.js [hunk 5]
```diff
@@ -46,21 +55,29 @@ const rule = (primary, secondaryOptions) => {
 			return;
 		}
 
+		/** @type {Map<string, number>} */ const insideFunctions = new Map(
+			Object.entries(secondaryOptions?.insideFunctions ?? {}),
+		);
```
Comment (ybiquitous) [inline lib/rules/number-max-precision/index.js:60]: [suggest] The following seems more consistent and readable a bit to me. Thoughts?

```suggestion
		/** @type {Map<string, number>} */
		const insideFunctions = new Map(Object.entries(secondaryOptions?.insideFunctions ?? {}));
```

**Review by ybiquitous — APPROVED** [review]
Thanks, LGTM! 👍🏼

File: lib/rules/number-max-precision/__tests__/index.js [hunk 6]
```diff
@@ -41,7 +41,7 @@ testRule({
 			code: "@IMPORT '1.123.css'",
 		},
 		{
-			code: 'a { background: url(1.123.jpg) }',
+			code: 'a { background: url(1.123.jpg) url("foo" 1.123.jpg) }',
 		},
 		{
 			code: 'a { my-string: "1.2345"; }',
```
Comment (Mouvedia) [inline lib/rules/number-max-precision/__tests__/index.js:47]: maybe replace this one by
```
font-family: "1.2345";
```
ref: https://drafts.csswg.org/css-fonts/#family-name-syntax

File: lib/rules/number-max-precision/__tests__/index.js [hunk 7]
```diff
@@ -41,7 +41,7 @@ testRule({
 			code: "@IMPORT '1.123.css'",
 		},
 		{
-			code: 'a { background: url(1.123.jpg) }',
+			code: 'a { background: url(1.123.jpg) url("foo" 1.123.jpg) }',
 		},
```
Comment (Mouvedia) [inline lib/rules/number-max-precision/__tests__/index.js:45]: suggestion

maybe add `image-set`?
https://developer.mozilla.org/en-US/docs/Web/CSS/image/image-set

File: lib/rules/number-max-precision/__tests__/index.js (hunk 6 above)
Comment (romainmenke) [reply lib/rules/number-max-precision/__tests__/index.js:47]: The description for this test is :
> ignore all strings rather than only in 'content'

The purpose seems to have been to test a non-specific property with a string value.
In contrast with a known property where string values are expected.

File: lib/rules/number-max-precision/__tests__/index.js (hunk 7 above)
Comment (romainmenke) [reply lib/rules/number-max-precision/__tests__/index.js:45]: The tests for `url` are a bit iffy in general.
I think these were intended to catch an implementation detail in `postcss-value-parser`. Not because anyone really wants to ignore floating point precision in numbers in `url`.

In `postcss-value-parser` `url(...)` is always parsed as a function and it's contents can appear to be numbers. e.g. `url(123.jpg)`

This however is incorrect and the CSS specification is very different.

Either `url(...)` is a single `url` token and it's value is a url or it is a function with a name of `url` and it contains a value list that starts with a string token.

- `url(123.jpg)` -> url token
- `url("123.jpg")` -> function with a name of url

Because `@csstools/css-tokenizer` and `@csstools/css-parser-algorithm` follow the CSS specification they will never produce a `url` function where the url value is parsed as a number.

-------

This is also why I had to add a [url modified](https://www.w3.org/TR/css-values-4/#url-modifiers) `url("foo" 1.123.jpg);`. Without this the test coverage wasn't 100%.

Currently no url modifiers exist in the CSS specification, so this isn't testing realistic CSS anymore.

-------

TL;DR;

Either we should remove the special handling of `url` or we should leave it as is.
I don't think we should add more tests for something that is actually an implementation detail in a parser that (as of this change) isn't used anymore :)

Comment (Mouvedia) [reply lib/rules/number-max-precision/__tests__/index.js:45]: suggestion

Add a concise one sentence _why_ comment above the `if` that ignores `url`?

Comment (romainmenke) [reply lib/rules/number-max-precision/__tests__/index.js:45]: I've added a note, please let me know if this is unclear in any way.

**Comment by romainmenke** [comment]
Thank you everyone for the thoughtful reviews 🙇
